Bob's Bookstore sells books to customers via an online platform. Most books at Bob's are animal themed. \
Information about the store's books can be found on their website, https://btech-data-analytics.github.io/bridgerland-technical-college/home.html.

We've been hired by Bob to perform an analysis on the books sold. \
It's important to note, Bob does not keep Excel or CSV files and instead simply adds and removes titles from the store's webpage as needed.

Below we will create a web scraper to gather the data from the 'bookstore' section of the Bob's Bookstore website into a pandas dataframe. \
Then answer the following questions: 
* Which author has the most books listed at Bob's Bookstore?
* Which is the most popular topic among books at Bob's Bookstore?
* Which topic of books is the most expensive, on average?
* Which topic of books has the most pages, on average?

In [2]:
#import statements
import pandas as pd
import requests
from bs4 import BeautifulSoup

In [58]:
response = requests.get('https://btech-data-analytics.github.io/bridgerland-technical-college/bookstore.html').text

In [59]:
soup = BeautifulSoup(response,'html.parser')

In [60]:
soup

<html lang="en">
<head>
<meta charset="utf-8"/>
<meta content="width=device-width, initial-scale=1.0" name="viewport"/>
<link href="https://www.w3schools.com/w3css/4/w3.css" rel="stylesheet"/>
<link href="style.css" rel="stylesheet"/>
<title>Bob's Bookstore</title>
</head>
<body>
<div class="w3-bar w3-blue-grey w3-large" style="padding: 0rem 8rem 0rem 8rem">
<a href="/bridgerland-technical-college/home.html"><img class="" src="bobs_bookstore_white.png" style="max-height: 2.8rem;"/></a>
<a class="w3-bar-item w3-button w3-right" href="/bridgerland-technical-college/bookstore.html">Store</a>
<a class="w3-bar-item w3-button w3-right" href="/bridgerland-technical-college/contact.html">Contact</a>
<a class="w3-bar-item w3-button w3-right" href="/bridgerland-technical-college/about.html">About</a>
</div>
<div class="w3-container" style="max-width: 1000px;margin: auto;">
<h1>Books for Sale</h1>
<table class="w3-table-all w3-hoverable w3-small">
<thead>
<tr>
<th>ISBN</th>
<th>Title</th>
<th>Aut

In [69]:
book_titles = []

book_rows = soup.find_all('tr', class_= 'book')

for book in book_rows: 
    cells = book.find_all('td')
    if len(cells) > 1: 
        title_text = cells[1].text.strip()
    book_titles.append(title_text)

print(book_titles)

['Whiskers of Wisdom: Tales from Feline Philosophers', "Purrfectly Pawesome: A Cat's Life", 'Cat Tales: Adventures in Whiskerland', 'The Enigmatic Paws: Mysteries of Meowville', 'Cats in Wonderland', 'Whisker Wisdom: Life Lessons from Feline Sages', 'Catnip Chronicles: A Purrfect Journey', 'Cat-astrophe: Tales of Misadventures', "The Cat's Whisker: A Feline Fantasy", 'Fur and Friendship: Stories of Feline Companions', 'Tails of Loyalty: Canine Chronicles', "Pawsitive Adventures: A Dog's Journey", 'Barking Wisdom: Lessons from Wise Canines', 'Dogged Determination: Stories of Resilient Pooches', 'The Bark Brigade: Canine Heroes Among Us']


In [81]:
ISBNs = []
book_titles = []
authors = []
languages = []
page_numbers = []
topics = []
prices = []

book_rows = soup.find_all('tr', class_= 'book')

for book in book_rows:
    cells = book.find_all('td')
    if len(cells) >=7:
        ISBNs.append(cells[0].text.strip())
        book_titles.append(cells[1].text.strip())
        authors.append(cells[2].text.strip())
        languages.append(cells[3].text.strip())
        page_numbers.append(cells[4].text.strip())
        topics.append(cells[5].text.strip())
        prices.append(cells[6].text.strip())
    
print(ISBNs)
print(book_titles)
print(authors)
print(language)
print(page_numbers)
print(topics)
print(prices)

['978-1234567890', '978-2345678901', '978-3456789012', '978-4567890123', '978-5678901234', '978-6789012345', '978-7890123456', '978-8901234567', '978-9012345678', '978-0123456789', '978-1234567890', '978-2345678901', '978-3456789012', '978-4567890123', '978-5678901234']
['Whiskers of Wisdom: Tales from Feline Philosophers', "Purrfectly Pawesome: A Cat's Life", 'Cat Tales: Adventures in Whiskerland', 'The Enigmatic Paws: Mysteries of Meowville', 'Cats in Wonderland', 'Whisker Wisdom: Life Lessons from Feline Sages', 'Catnip Chronicles: A Purrfect Journey', 'Cat-astrophe: Tales of Misadventures', "The Cat's Whisker: A Feline Fantasy", 'Fur and Friendship: Stories of Feline Companions', 'Tails of Loyalty: Canine Chronicles', "Pawsitive Adventures: A Dog's Journey", 'Barking Wisdom: Lessons from Wise Canines', 'Dogged Determination: Stories of Resilient Pooches', 'The Bark Brigade: Canine Heroes Among Us']
['Penelope Wainwright', 'Jasper Sterling', 'Penelope Wainwright', 'Maximilian Thorne

In [98]:
import pandas as pd
df = pd.DataFrame ({
    'ISBN' :ISBNs, 
    'Title' :book_titles,
    'Author' :authors,
    'Language' :language,
    'Pages' :page_numbers,
    'Topics' :topics,
    'Price':prices
})

df.describe()

ISBN                                              Title  \
count               15                                                 15   
unique              10                                                 15   
top     978-1234567890  Whiskers of Wisdom: Tales from Feline Philosop...   
freq                 2                                                  1   

                     Author Language Pages Topics   Price  
count                    15       15    15     15      15  
unique                   10        1     8      2      13  
top     Penelope Wainwright  English   192   Cats  $21.99  
freq                      4       15     3     10       2

In [99]:
df.head()

ISBN                                              Title  \
0  978-1234567890  Whiskers of Wisdom: Tales from Feline Philosop...   
1  978-2345678901                  Purrfectly Pawesome: A Cat's Life   
2  978-3456789012               Cat Tales: Adventures in Whiskerland   
3  978-4567890123         The Enigmatic Paws: Mysteries of Meowville   
4  978-5678901234                                 Cats in Wonderland   

                Author Language Pages Topics   Price  
0  Penelope Wainwright  English   256   Cats  $19.99  
1      Jasper Sterling  English   192   Cats  $15.99  
2  Penelope Wainwright  English   320   Cats  $21.99  
3    Maximilian Thorne  English   288   Cats  $17.99  
4   Isadora Harrington  English   224   Cats  $16.99

In [100]:
df['Topics'].value_counts()

Topics
Cats    10
Dogs     5
Name: count, dtype: int64

Based on the data above, Cats are the most popular topic.

In [101]:
df['Author'].value_counts()

Author
Penelope Wainwright    4
Jasper Sterling        2
Benjamin Barkley       2
Maximilian Thorne      1
Isadora Harrington     1
Celeste Nightshade     1
Seraphina Montague     1
Sophie Shepherd        1
Oliver Obedience       1
Ruby Ruffington        1
Name: count, dtype: int64

Based on the above data, Penelope Wainwright has the most books listed at Bob's Bookstore.

In [102]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   ISBN      15 non-null     object
 1   Title     15 non-null     object
 2   Author    15 non-null     object
 3   Language  15 non-null     object
 4   Pages     15 non-null     object
 5   Topics    15 non-null     object
 6   Price     15 non-null     object
dtypes: object(7)
memory usage: 972.0+ bytes


In [103]:
df['Price'] = df['Price'].str.replace('$', '', regex=False).astype(float)

In [105]:
df.groupby('Topics')['Price'].mean()

Topics
Cats    17.79
Dogs    26.59
Name: Price, dtype: float64

According to the data above, Books on the topic of dogs are more expensive with an average of 26.59. Compared to the cats average of 17.79

In [108]:
df['Pages'] = df['Pages'].astype(int)

In [109]:
df.groupby('Topics')['Pages'].mean()

Topics
Cats    238.4
Dogs    256.0
Name: Pages, dtype: float64

Books on the topic of dogs have more pages; 256 vs 238 (cats).

In [ ]:
#Below is data from the Home page instead of the store page. Not used in this analysis but helpful to see the work.

In [56]:
#create colum to determine most popular topic (extracted from titles)
def categories(title):
    title = title.lower()
    if 'rabbits' in title or 'hop' in title or 'carrot' in title:
        return 'Rabbits'
    elif 'horses' in title or 'neigh' in title:
        return 'Horses'
    elif 'cat' in title or 'cats' in title or 'whisker' in title or 'feline' in title: 
        return 'Cats'
    elif 'dog' in title or 'bark' in title or 'woof' in title:
        return 'Dogs'
    else:
        return 'Other'
        
df['Topic'] = df['Title'].apply(categories)

In [57]:
df['Topic'].value_counts()

Topic
Cats     7
Other    2
Name: count, dtype: int64

Based on the data above, Cats are the most popular topic.